From: Andrew Rybchenko <Andrew.Rybchenko@oktetlabs.ru>
Date: Wed, 15 Jun 2016 17:43:00 +0100
Subject: sfc: Move last mc_promisc flag to EF10 filter table state
Patch-mainline: v4.8-rc1
Git-commit: b071c3a222db465b32bb6f04ba07264a1556a17c
References: bsc#1017967 FATE#321663

It is used for EF10 only and logically belongs to EF10 filter table state.
It is OK that it is reset to false on filter table recreation since all
filters are removed on destruction.

Signed-off-by: Edward Cree <ecree@solarflare.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: David Chang <dchang@suse.com>
---
 drivers/net/ethernet/sfc/ef10.c       |    7 +++++--
 drivers/net/ethernet/sfc/net_driver.h |    2 --
 2 files changed, 5 insertions(+), 4 deletions(-)

--- a/drivers/net/ethernet/sfc/ef10.c
+++ b/drivers/net/ethernet/sfc/ef10.c
@@ -83,6 +83,8 @@ struct efx_ef10_filter_table {
 	u16 ucdef_id;
 	u16 bcast_id;
 	u16 mcdef_id;
+/* Whether in multicast promiscuous mode when last changed */
+	bool mc_promisc_last;
 };
 
 /* An arbitrary search limit for the software hash table */
@@ -3790,6 +3792,7 @@ static int efx_ef10_filter_table_probe(s
 	table->ucdef_id = EFX_EF10_FILTER_ID_INVALID;
 	table->bcast_id = EFX_EF10_FILTER_ID_INVALID;
 	table->mcdef_id = EFX_EF10_FILTER_ID_INVALID;
+	table->mc_promisc_last = false;
 
 	efx->filter_state = table;
 	init_waitqueue_head(&table->waitq);
@@ -4255,7 +4258,7 @@ static void efx_ef10_filter_sync_rx_mode
 	/* If changing promiscuous state with cascaded multicast filters, remove
 	 * old filters first, so that packets are dropped rather than duplicated
 	 */
-	if (nic_data->workaround_26807 && efx->mc_promisc != mc_promisc)
+	if (nic_data->workaround_26807 && table->mc_promisc_last != mc_promisc)
 		efx_ef10_filter_remove_old(efx);
 	if (mc_promisc) {
 		if (nic_data->workaround_26807) {
@@ -4290,7 +4293,7 @@ static void efx_ef10_filter_sync_rx_mode
 	}
 
 	efx_ef10_filter_remove_old(efx);
-	efx->mc_promisc = mc_promisc;
+	table->mc_promisc_last = mc_promisc;
 }
 
 static int efx_ef10_set_mac_address(struct efx_nic *efx)
--- a/drivers/net/ethernet/sfc/net_driver.h
+++ b/drivers/net/ethernet/sfc/net_driver.h
@@ -916,7 +916,6 @@ struct vfdi_status;
  * @stats_lock: Statistics update lock. Must be held when calling
  *	efx_nic_type::{update,start,stop}_stats.
  * @n_rx_noskb_drops: Count of RX packets dropped due to failure to allocate an skb
- * @mc_promisc: Whether in multicast promiscuous mode when last changed
  *
  * This is stored in the private area of the &struct net_device.
  */
@@ -1065,7 +1064,6 @@ struct efx_nic {
 	int last_irq_cpu;
 	spinlock_t stats_lock;
 	atomic_t n_rx_noskb_drops;
-	bool mc_promisc;
 };
 
 static inline int efx_dev_registered(struct efx_nic *efx)
